### DISCUSSION TITLE
Tournament client: Importing rounds and teams from tournament spreadsheet/external sources

### DISCUSSION THREAD

**Discussion by ILW8** [ppy/osu#24329] [Ideas]
Creating rounds and teams for a tournament is a time consuming process, despite all the required data to generate a list of teams and mappools for rounds being often readily available in the tournament's spreadsheet(s).

(see below for old proposed schema)
## Proposal
Spreadsheets can relatively easily generate JSON that is compatible with `osu.Game.Tournament.Models.TournamentTeam` and `osu.Game.Tournament.Models.TournamentRound` classes. I'd like to add a button that allows importing a TournamentTeam or a TournamentRound into the current tournament by allowing the user to paste a JSON into the lazer tournament client. 

This has the benefit of not needing to generate the full `bracket.json` or having the streamer go into `bracket.json` and manually fiddling with the `Rounds` and `Teams`, instead allowing the user to incrementally add teams/rounds as the tournament progresses through the UI.

_If_ `bracket.json` were to be split up into multiple files (mentioned in the below comments, I'm imagining `rounds.json`, `teams.json`, `bracket.json`, and `settings.json`), then it might make less sense to have UI elements for importing teams. That being said, inserting a round every week could still be easier by copy-pasting a JSON from spreadsheets into tourney client UI compared to saving a file and copying it into the tournament's folder.

https://github.com/ppy/osu/assets/1176059/f2e556bb-a1f4-4256-8417-ed2cf3bc0848

code for example shown is at https://github.com/ILW8/osu/tree/tourney-import-pools-and-teams

### Old proposal
<details>
  <summary>click me</summary>

I'd like to implement a tournament rounds and teams importer into the lazer tournament client. This is a request for comments on my plan so far.

In the tournament client's setup screen, I'd add "Import teams" and "Import rounds/mappools" buttons. These buttons will open a file browser to let the user provide a JSON file (or allow them to paste a JSON string) containing information about teams or rounds respectively in a standardised format. 

We've already seen spreadsheeters write formulas to generate JSONs about round and team data directly from within referee/mappooling sheets. I can see sheeter incorporate the generation of these JSONs into sheet templates, which will hopefully make importing teams and rounds into lazer a much less time-consuming process every week.

### Rounds list JSON schema
~~**edit**: these schemas should be updated to mirror `osu.Game.Tournament.Models.TournamentTeam` and `TournamentRound`~~

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Rounds list schema",
  "type": "array",
  "items": {
    "type": "object",
    "properties": {
      "round_name": {
        "type": "string"
      },
      "description": {
        "type": "string"
      },
      "start_time": {
        "type": "string"
      },
      "best_of": {
        "type": "integer",
        "minimum": 1
      },
      "beatmaps": {
        "type": "array",
        "items": {
          "type": "object",
          "properties": {
            "beatmap_id": {
              "type": "number"
            },
            "mods": {
              "type": "string"
            }
          },
          "required": [
            "beatmap_id",
            "mods"
          ]
        }
      }
    },
    "required": [
      "beatmaps"
    ]
  }
}
```
#### Example
```json
[
  {
    "start_time": "2023-07-27T16:00:00Z",
    "round_name": "Semifinals",
    "best_of": 11,
    "description": "Quarterfinals round for tournament XYZ",
    "beatmaps": [
      {
        "beatmap_id": 3707396,
        "mods": "NM"
      },
      {
        "beatmap_id": 271242,
        "mods": "NM"
      },
      {
        "beatmap_id": 87241,
        "mods": "HD"
      }
    ]
  },
  {
    "start_time": "2023-07-28T12:00:00Z",
    "round_name": "Finals",
    "best_of": 13,
    "description": "Finals round for tournament XYZ",
    "beatmaps": [
      {
        "beatmap_id": 1297812,
        "mods": "NM"
      },
      {
        "beatmap_id": 2197125,
        "mods": "NM"
      },
      {
        "beatmap_id": 1286732,
        "mods": "HD"
      },
      {
        "beatmap_id": 237519,
        "mods": "HR"
      }
    ]
  }
]
```
### Teams list JSON schema
```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Teams list schema",
  "type": "array",
  "items": {
    "type": "object",
    "properties": {
      "name": {
        "type": "string"
      },
      "acronym": {
        "type": "string"
      },
      "flag": {
        "type": "string"
      },
      "seed": {
        "type": "string"
      },
      "last_year_placement": {
        "type": "integer",
        "minimum": 0
      },
      "players": {
        "type": "array",
        "items": {
          "type": "integer",
          "minimum": 0
        }
      }
    },
    "required": [
      "players"
    ]
  }
}

```
#### Example
```json
[
  {
    "name": "Japan",
    "acronym": "JPN",
    "flag": "JP",
    "seed": "Low",
    "last_year_placement": 3,
    "players": [
      238192,
      129414,
      28314,
      288125
    ]
  },
  {
    "name": "United States",
    "acronym": "USA",
    "flag": "US",
    "seed": "",
    "last_year_placement": 1,
    "players": [
      182575,
      122815,
      175212,
      92471
    ]
  }
]
```
</details>

**Comment by peppy** [comment]
All world cups are already generating the bootstrap json and loading into osu. Are you just looking for a UI button so.. user don't have to copy into the tournament folder?

Or are you wanting to permanent json schemas so other apps implement this standard?

**Reply by ILW8** [reply]
The latter, a schema so community-run tournaments can implement the standard in their pooling/teams sheets

**Reply by peppy** [reply]
Then that's mostly just about documenting the format right? On osu!wiki or similar?

**Reply by ILW8** [reply]
Sorry I wasn't being very clear, the format should indeed be documented but the idea here is about adding UI buttons to import pools and teams from a json describing said pools and teams. 

I'm avoiding having the sheeters generate the whole `bracket.json` because that file includes settings, things typically not present in a tourney spreadsheet (like match node coordinates in bracket screen), or things that are less trivial to generate in terms of spreadsheet formulas (progressions).
The pools/rounds import button could be used every week to import the next round of the tournament without touching existing teams/matches/progressions, for example.

**Reply by peppy** [reply]
Maybe we should split the current single file into multiple files instead, hmm.

**Reply by ILW8** [reply]
I'll have a go implementing it and open a PR, I'll be using it for myself even if ultimately doesn't get merged. Having a working PoC might make it easier to visualize what I'm trying to achieve.
